KerberosTicketValidatorTests failed due to port already in use

The `KerberosTicketValidatorTests` suite failed in https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+master+g1gc/56/console

The repro command is:

```
./gradlew :x-pack:qa:evil-tests:test \
  -Dtests.seed=AAA9113F98B4F706 \
  -Dtests.class=org.elasticsearch.xpack.security.authc.kerberos.KerberosTicketValidatorTests \
  -Dtests.security.manager=false \
  -Dtests.jvm.argline="-XX:-UseConcMarkSweepGC -XX:+UseG1GC" \
  -Dtests.locale=en-US \
  -Dtests.timezone=Etc/UTC \
  -Dcompiler.java=10 \
  -Druntime.java=8
```

This doesn't reproduce locally.

The log suggests that the problem may be due to using a hardcoded port number that wasn't available in the test run that failed:

```
   > Throwable #1: java.security.PrivilegedActionException: org.apache.kerby.kerberos.kerb.KrbException: Failed to start KDC-Server
   > 	at __randomizedtesting.SeedInfo.seed([AAA9113F98B4F706:FB6252DAA8AC7C51]:0)
   > 	at java.security.AccessController.doPrivileged(Native Method)
   > 	at org.elasticsearch.xpack.security.authc.kerberos.SimpleKdcLdapServer.<init>(SimpleKdcLdapServer.java:86)
   > 	at org.elasticsearch.xpack.security.authc.kerberos.KerberosTestCase.startSimpleKdcLdapServer(KerberosTestCase.java:110)
   > 	at java.lang.Thread.run(Thread.java:748)
   > Caused by: org.apache.kerby.kerberos.kerb.KrbException: Failed to start KDC-Server
   > 	at org.apache.kerby.kerberos.kerb.server.impl.AbstractInternalKdcServer.start(AbstractInternalKdcServer.java:73)
   > 	at org.apache.kerby.kerberos.kerb.server.KdcServer.start(KdcServer.java:268)
   > 	at org.apache.kerby.kerberos.kerb.server.SimpleKdcServer.start(SimpleKdcServer.java:174)
   > 	at org.elasticsearch.xpack.security.authc.kerberos.SimpleKdcLdapServer.prepareKdcServerAndStart(SimpleKdcLdapServer.java:151)
   > 	at org.elasticsearch.xpack.security.authc.kerberos.SimpleKdcLdapServer.init(SimpleKdcLdapServer.java:104)
   > 	at org.elasticsearch.xpack.security.authc.kerberos.SimpleKdcLdapServer.access$000(SimpleKdcLdapServer.java:39)
   > 	at org.elasticsearch.xpack.security.authc.kerberos.SimpleKdcLdapServer$2.run(SimpleKdcLdapServer.java:89)
   > 	at org.elasticsearch.xpack.security.authc.kerberos.SimpleKdcLdapServer$2.run(SimpleKdcLdapServer.java:86)
   > 	... 39 more
   > Caused by: java.net.BindException: Address already in use
   > 	at sun.nio.ch.Net.bind0(Native Method)
   > 	at sun.nio.ch.Net.bind(Net.java:433)
   > 	at sun.nio.ch.DatagramChannelImpl.bind(DatagramChannelImpl.java:691)
   > 	at org.apache.kerby.kerberos.kerb.transport.KdcNetwork.listen(KdcNetwork.java:70)
   > 	at org.apache.kerby.kerberos.kerb.server.impl.DefaultInternalKdcServerImpl.doStart(DefaultInternalKdcServerImpl.java:67)
   > 	at org.apache.kerby.kerberos.kerb.server.impl.AbstractInternalKdcServer.start(AbstractInternalKdcServer.java:71)
   > 	... 46 more
```

I think this is another instance of this issue in `SimpleKdcLdapServerTests.testPrincipalCreationAndSearchOnLdap` - at least it looks like the KDC server failed to start in the same way. I'm not sure if that's really the root cause because the logs are really garbled below where this appears

https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+6.4+matrix-java-periodic/ES_BUILD_JAVA=java10,ES_RUNTIME_JAVA=java8fips,nodes=virtual&&linux/32/console

[build-64-32.txt.zip](https://github.com/elastic/elasticsearch/files/2540439/build-64-32.txt.zip)

```
ERROR   0.04s J2 | SimpleKdcLdapServerTests.testPrincipalCreationAndSearchOnLdap <<< FAILURES!
   > Throwable #1: java.security.PrivilegedActionException: org.apache.kerby.kerberos.kerb.KrbException: Failed to start KDC-Server
   >    at __randomizedtesting.SeedInfo.seed([5225DED06218B08D:15320F9392FA82F8]:0)
   >    at java.security.AccessController.doPrivileged(Native Method)
   >    at org.elasticsearch.xpack.security.authc.kerberos.SimpleKdcLdapServer.<init>(SimpleKdcLdapServer.java:86)
   >    at org.elasticsearch.xpack.security.authc.kerberos.KerberosTestCase.startSimpleKdcLdapServer(KerberosTestCase.java:110)
   >    at java.lang.Thread.run(Thread.java:748)
   > Caused by: org.apache.kerby.kerberos.kerb.KrbException: Failed to start KDC-Server
   >    at org.apache.kerby.kerberos.kerb.server.impl.AbstractInternalKdcServer.start(AbstractInternalKdcServer.java:73)
   >    at org.apache.kerby.kerberos.kerb.server.KdcServer.start(KdcServer.java:268)
   >    at org.apache.kerby.kerberos.kerb.server.SimpleKdcServer.start(SimpleKdcServer.java:174)
   >    at org.elasticsearch.xpack.security.authc.kerberos.SimpleKdcLdapServer.prepareKdcServerAndStart(SimpleKdcLdapServer.java:151)
   >    at org.elasticsearch.xpack.security.authc.kerberos.SimpleKdcLdapServer.init(SimpleKdcLdapServer.java:104)
   >    at org.elasticsearch.xpack.security.authc.kerberos.SimpleKdcLdapServer.access$000(SimpleKdcLdapServer.java:39)
   >    at org.elasticsearch.xpack.security.authc.kerberos.SimpleKdcLdapServer$2.run(SimpleKdcLdapServer.java:89)
   >    at org.elasticsearch.xpack.security.authc.kerberos.SimpleKdcLdapServer$2.run(SimpleKdcLdapServer.java:86)
   >    ... 39 more
   > Caused by: java.net.BindException: Address already in use
   >    at sun.nio.ch.Net.bind0(Native Method)
   >    at sun.nio.ch.Net.bind(Net.java:433)
   >    at sun.nio.ch.DatagramChannelImpl.bind(DatagramChannelImpl.java:691)
   >    at org.apache.kerby.kerberos.kerb.transport.KdcNetwork.listen(KdcNetwork.java:70)
   >    at org.apache.kerby.kerberos.kerb.server.impl.DefaultInternalKdcServerImpl.doStart(DefaultInternalKdcServerImpl.java:67)
   >    at org.apache.kerby.kerberos.kerb.server.impl.AbstractInternalKdcServer.start(AbstractInternalKdcServer.java:71)
   >    ... 46 more
```

Another one in `KerberosTicketValidatorTests.testValidKebrerosTicket`

```
ERROR   0.02s J1 | KerberosTicketValidatorTests.testValidKebrerosTicket <<< FAILURES!
   > Throwable #1: java.security.PrivilegedActionException: org.apache.kerby.kerberos.kerb.KrbException: Failed to start KDC-Server
   >    at __randomizedtesting.SeedInfo.seed([FE44AA7E6093930C:70F30192B5DCCA21]:0)
   >    at java.security.AccessController.doPrivileged(Native Method)
   >    at org.elasticsearch.xpack.security.authc.kerberos.SimpleKdcLdapServer.<init>(SimpleKdcLdapServer.java:86)
   >    at org.elasticsearch.xpack.security.authc.kerberos.KerberosTestCase.startSimpleKdcLdapServer(KerberosTestCase.java:111)
   >    at java.lang.Thread.run(Thread.java:748)
   > Caused by: org.apache.kerby.kerberos.kerb.KrbException: Failed to start KDC-Server
   >    at org.apache.kerby.kerberos.kerb.server.impl.AbstractInternalKdcServer.start(AbstractInternalKdcServer.java:73)
   >    at org.apache.kerby.kerberos.kerb.server.KdcServer.start(KdcServer.java:268)
   >    at org.apache.kerby.kerberos.kerb.server.SimpleKdcServer.start(SimpleKdcServer.java:174)
   >    at org.elasticsearch.xpack.security.authc.kerberos.SimpleKdcLdapServer.prepareKdcServerAndStart(SimpleKdcLdapServer.java:151)
   >    at org.elasticsearch.xpack.security.authc.kerberos.SimpleKdcLdapServer.init(SimpleKdcLdapServer.java:104)
   >    at org.elasticsearch.xpack.security.authc.kerberos.SimpleKdcLdapServer.access$000(SimpleKdcLdapServer.java:39)
   >    at org.elasticsearch.xpack.security.authc.kerberos.SimpleKdcLdapServer$2.run(SimpleKdcLdapServer.java:89)
   >    at org.elasticsearch.xpack.security.authc.kerberos.SimpleKdcLdapServer$2.run(SimpleKdcLdapServer.java:86)
   >    ... 39 more
   > Caused by: java.net.BindException: Address already in use
   >    at sun.nio.ch.Net.bind0(Native Method)
   >    at sun.nio.ch.Net.bind(Net.java:433)
   >    at sun.nio.ch.DatagramChannelImpl.bind(DatagramChannelImpl.java:691)
   >    at org.apache.kerby.kerberos.kerb.transport.KdcNetwork.listen(KdcNetwork.java:70)
   >    at org.apache.kerby.kerberos.kerb.server.impl.DefaultInternalKdcServerImpl.doStart(DefaultInternalKdcServerImpl.java:67)
   >    at org.apache.kerby.kerberos.kerb.server.impl.AbstractInternalKdcServer.start(AbstractInternalKdcServer.java:71)
   >    ... 46 more
```

https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+6.x+intake/430/console

[build-430.txt.zip](https://github.com/elastic/elasticsearch/files/2641209/build-430.txt.zip)

--- a/x-pack/qa/evil-tests/src/test/java/org/elasticsearch/xpack/security/authc/kerberos/SimpleKdcLdapServer.java
+++ b/x-pack/qa/evil-tests/src/test/java/org/elasticsearch/xpack/security/authc/kerberos/SimpleKdcLdapServer.java
@@ -238,13 +238,13 @@ public class SimpleKdcLdapServer {
         if (transport != null && transport.trim().equalsIgnoreCase("tcp")) {
             try (ServerSocket serverSocket = ServerSocketFactory.getDefault().createServerSocket(port, 1,
                     InetAddress.getByName("127.0.0.1"))) {
-                return true;
+                return serverSocket.isBound();
             } catch (Exception ex) {
                 return false;
             }
         } else if (transport != null && transport.trim().equalsIgnoreCase("udp")) {
             try (DatagramSocket socket = new DatagramSocket(port, InetAddress.getByName("127.0.0.1"))) {
-                return true;
+                return socket.isBound();
             } catch (Exception ex) {
                 return false;
             }

--- a/x-pack/qa/evil-tests/src/test/java/org/elasticsearch/xpack/security/authc/kerberos/SimpleKdcLdapServer.java
+++ b/x-pack/qa/evil-tests/src/test/java/org/elasticsearch/xpack/security/authc/kerberos/SimpleKdcLdapServer.java
@@ -22,6 +22,9 @@ import org.elasticsearch.common.unit.TimeValue;
 import org.elasticsearch.test.ESTestCase;
 
 import java.io.IOException;
+import java.net.DatagramSocket;
+import java.net.InetAddress;
+import java.net.ServerSocket;
 import java.nio.charset.StandardCharsets;
 import java.nio.file.Files;
 import java.nio.file.Path;
@@ -30,6 +33,8 @@ import java.security.PrivilegedActionException;
 import java.security.PrivilegedExceptionAction;
 import java.util.Locale;
 import java.util.concurrent.TimeUnit;
+
+import javax.net.ServerSocketFactory;
 
 /**
  * Utility wrapper around Apache {@link SimpleKdcServer} backed by Unboundid
@@ -129,6 +134,14 @@ public class SimpleKdcLdapServer {
         simpleKdc.setKdcRealm(realm);
         if (kdcPort == 0) {
             kdcPort = NetworkUtil.getServerPort();
+            int tries = 0;
+            while (isPortAvailable(kdcPort, transport) == false) {
+                if (tries > 100) {
+                    throw new IllegalArgumentException("Could not find available port for KDC server");
+                }
+                kdcPort = NetworkUtil.getServerPort();
+                tries++;
+            }
         }
         if (transport != null) {
             if (transport.trim().equals("TCP")) {
@@ -221,4 +234,21 @@ public class SimpleKdcLdapServer {
         logger.info("SimpleKdcServer stoppped.");
     }
 
+    private boolean isPortAvailable(int port, String transport) {
+        if (transport != null && transport.trim().equalsIgnoreCase("tcp")) {
+            try (ServerSocket serverSocket = ServerSocketFactory.getDefault().createServerSocket(port, 1,
+                    InetAddress.getByName("127.0.0.1"))) {
+                return true;
+            } catch (Exception ex) {
+                return false;
+            }
+        } else if (transport != null && transport.trim().equalsIgnoreCase("udp")) {
+            try (DatagramSocket socket = new DatagramSocket(port, InetAddress.getByName("127.0.0.1"))) {
+                return true;
+            } catch (Exception ex) {
+                return false;
+            }
+        }
+        return false;
+    }
 }

--- a/x-pack/qa/evil-tests/src/test/java/org/elasticsearch/xpack/security/authc/kerberos/SimpleKdcLdapServer.java
+++ b/x-pack/qa/evil-tests/src/test/java/org/elasticsearch/xpack/security/authc/kerberos/SimpleKdcLdapServer.java
@@ -135,10 +135,10 @@ public class SimpleKdcLdapServer {
             if (kdcPort == 0) {
                 kdcPort = getServerPort(transport);
             }
-            if (transport.trim().equals("TCP")) {
+            if (transport.trim().equalsIgnoreCase("TCP")) {
                 simpleKdc.setKdcTcpPort(kdcPort);
                 simpleKdc.setAllowUdp(false);
-            } else if (transport.trim().equals("UDP")) {
+            } else if (transport.trim().equalsIgnoreCase("UDP")) {
                 simpleKdc.setKdcUdpPort(kdcPort);
                 simpleKdc.setAllowTcp(false);
             } else {
@@ -226,18 +226,18 @@ public class SimpleKdcLdapServer {
     }
 
     private static int getServerPort(String transport) {
-        if (transport != null && transport.trim().equalsIgnoreCase("tcp")) {
+        if (transport != null && transport.trim().equalsIgnoreCase("TCP")) {
             try (ServerSocket serverSocket = ServerSocketFactory.getDefault().createServerSocket(0, 1,
                     InetAddress.getByName("127.0.0.1"))) {
                 return serverSocket.getLocalPort();
             } catch (Exception ex) {
-                throw new RuntimeException("Failed to get a tcp server socket point");
+                throw new RuntimeException("Failed to get a TCP server socket point");
             }
-        } else if (transport != null && transport.trim().equalsIgnoreCase("udp")) {
+        } else if (transport != null && transport.trim().equalsIgnoreCase("UDP")) {
             try (DatagramSocket socket = new DatagramSocket(0, InetAddress.getByName("127.0.0.1"))) {
                 return socket.getLocalPort();
             } catch (Exception ex) {
-                throw new RuntimeException("Failed to get a udp server socket point");
+                throw new RuntimeException("Failed to get a UDP server socket point");
             }
         }
         throw new IllegalArgumentException("Invalid transport: " + transport);

--- a/x-pack/qa/evil-tests/src/test/java/org/elasticsearch/xpack/security/authc/kerberos/SimpleKdcLdapServer.java
+++ b/x-pack/qa/evil-tests/src/test/java/org/elasticsearch/xpack/security/authc/kerberos/SimpleKdcLdapServer.java
@@ -233,6 +233,7 @@ public class SimpleKdcLdapServer {
                 throw new IllegalArgumentException("Could not find available port for KDC server");
             }
             port = ESTestCase.randomIntBetween(1024, 65535);
+            tries++;
         } while (isPortAvailable(port, transport) == false);
         return port;
     }

--- a/x-pack/qa/evil-tests/src/test/java/org/elasticsearch/xpack/security/authc/kerberos/SimpleKdcLdapServer.java
+++ b/x-pack/qa/evil-tests/src/test/java/org/elasticsearch/xpack/security/authc/kerberos/SimpleKdcLdapServer.java
@@ -13,7 +13,6 @@ import org.apache.kerby.kerberos.kerb.KrbException;
 import org.apache.kerby.kerberos.kerb.client.KrbConfig;
 import org.apache.kerby.kerberos.kerb.server.KdcConfigKey;
 import org.apache.kerby.kerberos.kerb.server.SimpleKdcServer;
-import org.apache.kerby.util.NetworkUtil;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 import org.elasticsearch.ExceptionsHelper;
@@ -132,18 +131,10 @@ public class SimpleKdcLdapServer {
         simpleKdc.setWorkDir(workDir.toFile());
         simpleKdc.setKdcHost(host);
         simpleKdc.setKdcRealm(realm);
-        if (kdcPort == 0) {
-            kdcPort = NetworkUtil.getServerPort();
-            int tries = 0;
-            while (isPortAvailable(kdcPort, transport) == false) {
-                if (tries > 100) {
-                    throw new IllegalArgumentException("Could not find available port for KDC server");
-                }
-                kdcPort = NetworkUtil.getServerPort();
-                tries++;
-            }
-        }
         if (transport != null) {
+            if (kdcPort == 0) {
+                kdcPort = randomServerPort(transport);
+            }
             if (transport.trim().equals("TCP")) {
                 simpleKdc.setKdcTcpPort(kdcPort);
                 simpleKdc.setAllowUdp(false);
@@ -234,7 +225,19 @@ public class SimpleKdcLdapServer {
         logger.info("SimpleKdcServer stoppped.");
     }
 
-    private boolean isPortAvailable(int port, String transport) {
+    private static int randomServerPort(String transport) {
+        int tries = 0;
+        int port = -1;
+        do  {
+            if (tries > 100) {
+                throw new IllegalArgumentException("Could not find available port for KDC server");
+            }
+            port = ESTestCase.randomIntBetween(1024, 65535);
+        } while (isPortAvailable(port, transport) == false);
+        return port;
+    }
+
+    private static boolean isPortAvailable(int port, String transport) {
         if (transport != null && transport.trim().equalsIgnoreCase("tcp")) {
             try (ServerSocket serverSocket = ServerSocketFactory.getDefault().createServerSocket(port, 1,
                     InetAddress.getByName("127.0.0.1"))) {

--- a/x-pack/qa/evil-tests/src/test/java/org/elasticsearch/xpack/security/authc/kerberos/SimpleKdcLdapServer.java
+++ b/x-pack/qa/evil-tests/src/test/java/org/elasticsearch/xpack/security/authc/kerberos/SimpleKdcLdapServer.java
@@ -133,7 +133,7 @@ public class SimpleKdcLdapServer {
         simpleKdc.setKdcRealm(realm);
         if (transport != null) {
             if (kdcPort == 0) {
-                kdcPort = randomServerPort(transport);
+                kdcPort = getServerPort(transport);
             }
             if (transport.trim().equals("TCP")) {
                 simpleKdc.setKdcTcpPort(kdcPort);
@@ -225,34 +225,21 @@ public class SimpleKdcLdapServer {
         logger.info("SimpleKdcServer stoppped.");
     }
 
-    private static int randomServerPort(String transport) {
-        int tries = 0;
-        int port = -1;
-        do  {
-            if (tries > 100) {
-                throw new IllegalArgumentException("Could not find available port for KDC server");
-            }
-            port = ESTestCase.randomIntBetween(1024, 65535);
-            tries++;
-        } while (isPortAvailable(port, transport) == false);
-        return port;
-    }
-
-    private static boolean isPortAvailable(int port, String transport) {
+    private static int getServerPort(String transport) {
         if (transport != null && transport.trim().equalsIgnoreCase("tcp")) {
-            try (ServerSocket serverSocket = ServerSocketFactory.getDefault().createServerSocket(port, 1,
+            try (ServerSocket serverSocket = ServerSocketFactory.getDefault().createServerSocket(0, 1,
                     InetAddress.getByName("127.0.0.1"))) {
-                return serverSocket.isBound();
+                return serverSocket.getLocalPort();
             } catch (Exception ex) {
-                return false;
+                throw new RuntimeException("Failed to get a tcp server socket point");
             }
         } else if (transport != null && transport.trim().equalsIgnoreCase("udp")) {
-            try (DatagramSocket socket = new DatagramSocket(port, InetAddress.getByName("127.0.0.1"))) {
-                return socket.isBound();
+            try (DatagramSocket socket = new DatagramSocket(0, InetAddress.getByName("127.0.0.1"))) {
+                return socket.getLocalPort();
             } catch (Exception ex) {
-                return false;
+                throw new RuntimeException("Failed to get a udp server socket point");
             }
         }
-        return false;
+        throw new IllegalArgumentException("Invalid transport: " + transport);
     }
 }

